src/main/java/com/powsybl/openloadflow/sensi/DcSensitivityAnalysis.java

```diff
@@ -940,76 +548,56 @@ public void analyse(Network network, List<PropagatedContingency> contingencies,
-
-                // process contingencies with connectivity break
-                for (ConnectivityAnalysisResult connectivityAnalysisResult : connectivityAnalysisResults) {
-                    processContingenciesBreakingConnectivity(connectivityAnalysisResult, loadFlowContext, lfParameters, lfParametersExt,
-                            validFactorHolder, factorGroups, participatingElements,
-                            contingencyElementByBranch, flowStates, factorsStates, contingenciesStates, resultWriter, reportNode);
-                }
+                // engine to compute pre- and post-contingency states for sensitivities calculations
+                WoodburyEngine engine = new WoodburyEngine();
+
+                // pre-contingency injection/flow rhs
+                DenseMatrix injectionRhs = getPreContingencyInjectionRhs(loadFlowContext, factorGroups, participatingElements);
+                DenseMatrix flowsRhs = getPreContingencyFlowRhs(loadFlowContext, participatingElements, new DisabledNetwork());
+
+                // compute states with +1 -1 to model the contingencies and run connectivity analysis
+                ConnectivityBreakAnalysis.ConnectivityBreakAnalysisResults connectivityData = ConnectivityBreakAnalysis.run(loadFlowContext, contingencies);
+
+                // storage of modifications that must be applied on rhs members, due to GLSK and/or slack bus participation
+                WoodburyEngineRhsModifications woodburyEngineRhsModifications = new WoodburyEngineRhsModifications();
+
+                // storage of disabled network by propagated contingencies, for sensitivity calculation
+                HashMap<PropagatedContingency, DisabledNetwork> disabledNetworkByPropagatedContingency = new HashMap<>();
+
+                // compute rhs modifications for contingencies breaking connectivity
+                buildRhsModificationsForContingenciesBreakingConnectivity(loadFlowContext, lfParametersExt, connectivityData, factorGroups, participatingElements,
+                        woodburyEngineRhsModifications, disabledNetworkByPropagatedContingency, resultWriter);
+
+                // compute rhs modifications for contingencies with no connectivity break
+                buildRhsModificationsForContingencies(loadFlowContext, lfParametersExt, factorGroups, connectivityData.nonBreakingConnectivityContingencies(), participatingElements,
+                        woodburyEngineRhsModifications, disabledNetworkByPropagatedContingency, Collections.emptySet(), Collections.emptySet(),
+                        connectivityData.contingencyElementByBranch(), Collections.emptySet(), resultWriter);
+
+                // compute the pre- and post-contingency states using Woodbury equality
+                WoodburyEngineResult woodburyResult = engine.run(loadFlowContext, flowsRhs, injectionRhs, woodburyEngineRhsModifications, connectivityData, reportNode);
```

Comment:
You need to keep a bunch of dense matrix in memory here. Can this create a memory problem for a large number for contingencies ? Is the memory comparable to mem size of contingencies or can it be much larger ?
If much larger, maybe the run should be batched by a certain amount of contingencies (ex 1000) to keep memory requirements controlled (same would apply for security analysis).

Reply:
I close this problem. I've been told that the memory problem exists but it can be considered orthogonal to this refactoring effort.